File: magellan-library/src/main/java/com/wealthfront/magellan/navigation/NavigableListener.java
package com.wealthfront.magellan.navigation;

import android.content.Context;

import com.wealthfront.magellan.lifecycle.LifecycleAware;

import org.jetbrains.annotations.NotNull;

import androidx.annotation.CallSuper;

public interface NavigableListener extends LifecycleAware {

  default void onNavigableShown(@NotNull NavigableCompat navigable) {}

  default void onNavigableHidden(@NotNull NavigableCompat navigable) {}

  default void beforeNavigation() {}

  default void afterNavigation() {}

  @Override
  @CallSuper
  default void create(@NotNull Context context) {
    NavigationPropagator.addNavigableListener(this);
  }

  @Override
  @CallSuper
  default void destroy(@NotNull Context context) {
    NavigationPropagator.removeNavigableListener(this);
  }
}

File: magellan-library/src/main/java/com/wealthfront/magellan/navigation/NavigableListener.kt
package com.wealthfront.magellan.navigation

import android.content.Context
import com.wealthfront.magellan.lifecycle.LifecycleAware

interface NavigableListener : LifecycleAware {

  @JvmDefault
  fun onNavigableShown(navigable: NavigableCompat) {}

  @JvmDefault
  fun onNavigableHidden(navigable: NavigableCompat) {}

  @JvmDefault
  fun onNavigate() {}

  @JvmDefault
  override fun create(context: Context) {
    NavigationPropagator.addNavigableListener(this)
  }

  @JvmDefault
  override fun destroy(context: Context) {
    NavigationPropagator.removeNavigableListener(this)
  }
}
